fix: check the output type of ode_func

mirko-m opened the pull request:
This PR addresses the issue discussed here https://discourse.pymc.io/t/return-value-for-2n-dimensional-ode-system/4141?u=mirko-m .

When estimating the parameters of a system of ODEs we need to define a a function called `ode_func` which returns the right hand side of the ODE system. It is not clear what the return type of `ode_func` is. In some of the examples it is a list, but for some ODE systems the user may prefer to return a  `TensorVariable` or a `numpy.ndarray`. The changes in this PR check for the return type of `ode_func` and then convert to a `TensorVariable`.

I am not sure if using `at.shared()` is the best method for casting a`numpy.ndarray`  to a `TensorVariable` .

I do not have a test for the changes I made, but it should be possible to adapt one of the existing examples. Let me know if you need me to do this.

Commit:
fix: check the output type of ode_func

canyon289 commented:
@mirko-m thanks for making a contribution. As you noted please make a test. I'll defer your question about `at.shared` to one of the inhouse Aesara experts!

ricardoV94 commented:
I am not familiar with ODE module but I don't think `np.ndarray` is a ever a valid output since the lines immediately below are trying to extract gradients from it.

mirko-m commented:
Thanks for your replies. My changes actually broke some of the existing tests. I am trying to debug this, but in my development environment the `import pymc` statement takes a lot of time to execute (several minutes). This makes the debugging process very slow? Do you happen to know a fix or workaround for this?

I installed the `main` branch on Ubuntu-20.04 running under WSL using
```bash
pip install -r requirements-dev.txt
pip install -e .
```

Commit:
fix: case where TensorVariable is a scalar

Commit:
Add test case where the output of ode_func is a TensorVariable

Commit:
Raise TypeError when output of ode_func is np.ndarray

mirko-m commented:
I made the following changes:

- Fix the case where `ode_func` returns a scalar `TensorVariable`.
- Add a test for the case where `ode_func` returns a vector `TensorVariable`.
- Raise a `TypeError` when `ode_func` returns a np.ndarray.
- Add a test checking that the `TypeError` is raised.

Do we need to add more tests for the case where the output of` `ode_func` is a `TensorVariable`? And do we want to put them in a separate class or leave them in `TestSensitivityInitialCondition`?

michaelosthege commented on pymc/ode/utils.py:
```diff
@@ -103,10 +103,21 @@ def augment_system(ode_func, n_states, n_theta):
+            # Convert yhat from scalar to array
+            t_yhat = at.stack((yhat,), axis=0)
+        else:
+            t_yhat = yhat
```
```suggestion
        elif yhat.ndim == 2:
            t_yhat = yhat
        else:
            raise ValueError(f"The odefunc returned a {yhat.ndim}-dimensional tensor, but 0 or 1 dimensions were expected.")
```

michaelosthege reviewed (commented):
Looks good. I'd just like to cover this additional edge case with an informative error.

michaelosthege committed:
Raise error on unexpected tensor dimensionality

michaelosthege commented:
FYI: I fixed the pre-commit.
If all tests go ✔ this should be good to go.

ricardoV94 commented on pymc/ode/utils.py:
```diff
@@ -103,10 +103,25 @@ def augment_system(ode_func, n_states, n_theta):
+            raise ValueError(
+                f"The odefunc returned a {yhat.ndim}-dimensional tensor, but 0 or 1 dimensions were expected."
+            )
+    elif isinstance(yhat, np.ndarray):
```
This check is a bit arbitrary, no? Was the original issue related to this?

Otherwise why aren't we checking if the tuple/list are not composed of ndarrays as well?

michaelosthege replied:
On a second though, yes.
Maybe simply doing `at.atleast_1d(yhat)` and then checking `ndim==1`?
Do we even need to anticipate the `ndarray` case?

mirko-m replied:
I think it is possible that `ode_func` returns an `ndarray` . This is for example what was tried here https://discourse.pymc.io/t/return-value-for-2n-dimensional-ode-system/4141/3 . It looks like in this case `np.concatenate` threw an error, but maybe there are other operations which will cast a `TensorVariable` to an `ndarray` without throwing an error.

ricardoV94 replied:
That's not possible

mirko-m replied:
I removed the check for `np.ndarray` and I made the code more concise by using `at.atleast_1d` as suggested. Please let me know if there is anything else that needs to be done before this PR is good to go.

michaelosthege replied:
Thanks! We're working to fix the CI on main and once that's done we can rebase your branch so we can have a green CI before merging

Commit:
remove check for ndarray

Commit:
be more concise by using at.atleast_1d

ricardoV94 commented on pymc/ode/utils.py:
```diff
@@ -103,10 +103,17 @@ def augment_system(ode_func, n_states, n_theta):
-    t_yhat = at.stack(yhat, axis=0)
+    if isinstance(yhat, at.TensorVariable):
+        t_yhat = at.atleast_1d(yhat)
+        if t_yhat.ndim > 1:
```
Does this check also make sense in the `else` branch? If so, you could move it below the `if/else` statements.

ricardoV94 replied:
Also, can you add a test for it?

mirko-m replied:
I moved the check so that it also applies to the else branch.

I also created two tests which check that the `ValueError` is thrown.

Commit:
Check ndim in all cases

Commit:
tests for the ValueError when ndim is wrong

michaelosthege commented on pymc/tests/test_ode.py:
```diff
@@ -267,6 +267,22 @@ def test_number_of_params(self):
+        return [[s,s], [s,s]]
+
+    def test_list_shape(self):
+        with pytest.raises(ValueError):
```
```suggestion
        with pytest.raises(ValueError, match="returned a 2-dimensional tensor"):
```

michaelosthege commented on pymc/tests/test_ode.py:
```diff
@@ -267,6 +267,22 @@ def test_number_of_params(self):
+        return at.stack((s, s, s, s)).reshape((2,2))
+
+    def test_tensor_shape(self):
+        with pytest.raises(ValueError):
```
```suggestion
        with pytest.raises(ValueError, match="returned a 2-dimensional tensor"):
```

michaelosthege reviewed (commented):
Let's see what the CI tests say 👍

mirko-m committed:
Use match when testing for shape of list

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

mirko-m committed:
Use match when testing for shape of tensor

Co-authored-by: Michael Osthege <michael.osthege@outlook.com>

mirko-m commented:
After accepting the suggested changes it looks to me like the tests that I wrote are catching the wrong ValueError. I will fix that.

Commit:
fix tests for ndim

Commit:
add matches to error checks in test_ode.py

mirko-m commented:
I added two more commits. The first commit fixes the new tests that I had written to check that when `yhat` is 2-dimensional a `ValueError` is thrown (`TestErrors.test_tensor_shape` and `TestErrors.test_list_shape`).

The second commit adds the `match` keyword from `pytest.raises` to the other tests inside the `TestErrors` class. I think that this is useful, but I am not sure if you want to merge it as part of this pull request or a separate pull request, because perhaps it it off topic. That is why I created a second commit for this change, so that you can easily ignore it.

When fixing the new tests (`TestErrors.test_tensor_shape` and `TestErrors.test_list_shape`) I noticed that `TestErrors.system`, `TestErrors.system_2d_tensor` and `TestErrors.system_2d_list` are not really instance methods of the class and therefore cannot be called as e.g. `self.system` . Instead in order to get the new tests to work I had to make the calls as `TestErrors.system_2d_tensor` and `TestErrors.system_2d_list`, respectively. For the other tests that are part of the `TestErrors` class this does not matter so much, because the errors that `pytest` expects are raised before `TestErrors.system` is called, but I changed the calls from `self.system` to `TestErrors.system` anyways because I think that this the correct way of calling it. Let me know if you disagree or if there is a better way of calling these methods/functions.

michaelosthege commented on pymc/tests/test_ode.py:
```diff
@@ -218,35 +237,57 @@ def system(y, t, p):
+
+    def test_list_shape(self):
+        with pytest.raises(ValueError, match="returned a 2-dimensional tensor"):
+            DifferentialEquation(func=TestErrors.system_2d_list, t0=0, times=self.times, n_states=4, n_theta=1)
```
If the `system_*` functions are just used inside one test each, better define them inside the test methods.

Otherwise decorate them with `@staticmethod` and continue to reference them like `TestErrors.system_*`

mirko-m replied:
The `system_2d_*` functions are only used for a single test, so I moved the definition inside those tests.

The function called `system` is used in multiple tests and I therefore decorated it with `@staticmethod`. This means that on line 238 I need to change the call to `func=system.__func__` because `system` is not callable. If instead I try to use `func=TestErrors.system` I get a `NameError`, because `TestErrors` is not defined.

Commit:
improve how we deal with staticmethods in TestErrors

Commit:
remove print statement from debugging

Commit:
fix pre-commit

michaelosthege commented on pymc/tests/test_ode.py:
```diff
@@ -209,44 +228,85 @@ def system_1(y, t, p):
     times = np.arange(0, 9)
 
-    ode_model = DifferentialEquation(func=system, t0=0, times=times, n_states=1, n_theta=1)
+    ode_model = DifferentialEquation(func=system.__func__, t0=0, times=times, n_states=1, n_theta=1)
```
This would be typical for a `setup_method`:

```python
def setup_method(self, method):
    def system(y, t, p):
        return np.exp(-t) - p[0] * y[0]
    self.times = np.arange(0, 9)
    self.ode_model = DifferentialEquation(func=system, t0=0, times=self.times, n_states=1, n_theta=1)
```
Also see https://docs.pytest.org/en/latest/how-to/xunit_setup.html

mirko-m replied:
I implemented the setup method and I will try to rebase onto main next.

mirko-m replied:
Looks like the rebase fixed the CI pipeline. Let me know what you think.

michaelosthege reviewed (commented):
Hi @mirko-m,
sorry that this is taking so long.
I think you'll need to rebase onto the main branch, otherwise the CI pipeline will fail, because of installation dependencies.
Let me know if you need help with it

Commit:
use setup_method in TestErrors

Commit:
fix pre-commit

michaelosthege commented on pymc/ode/utils.py:
```diff
@@ -103,10 +103,17 @@ def augment_system(ode_func, n_states, n_theta):
+    else:
+        # Stack the results of the ode_func into a single tensor variable
+        if not isinstance(yhat, (list, tuple)):
+            yhat = (yhat,)
```
The CI says this is not covered by the tests?
I think this line could become an exception: Unexpected type returned by the ODE function.
And you could test it by returning a dictionary or set of ODEs instead of a tensor/list/tuple.

FYI you can generate the coverage report locally by passing `--cov=pymc --cov-report=term-missing`. Needs "pip install pytest-cov" and can be combined with running just a specific test file.

I'll approve the CI run as soon as I notice a commit. Let's get this merged on Saturday, because I want to do another typing related PR this weekend. I can also take over from here if you want to

mirko-m replied:
I was away over the weekend, but I made the changes that you suggested just now. Hopefully the CI still works.

Commit:
raise a TypeEror when ode_func returns an unexpected type

michaelosthege commented:
Thanks, @mirko-m !

mirko-m commented:
Thanks for guiding me through the process and all your suggestions.
